back/app/models/initiative.rb

```diff
@@ -113,7 +113,12 @@ class Initiative < ApplicationRecord
   def cosponsor_ids=(ids)
     return unless ids
 
-    super(ids.uniq)
+    ids = ids.uniq
+    current_ids = cosponsors.pluck(:id).uniq
+    return if current_ids.sort == ids.sort
+
+    cosponsors_initiatives.where.not(user_id: ids).destroy_all
+    (ids - current_ids).each { |id| cosponsors_initiatives.create(user_id: id) }
```

Comment:
Using the more verbose version, since it uses `destroy`, thus enabling the `before_destroy :remove_notifications` hook in the `cosponsors_initiative` model to do its thing, which in turn means deleting `cosponsors_initiatives` records with an associated `notification` does not raise a `PG::ForeignKeyViolation`.